ui/src/app/pages/admin/admin-institution-edit.tsx

```diff
@@ -0,0 +1,411 @@
+import * as React from 'react';
+import * as validate from 'validate.js';
+
+const styles = reactStyles({
+  label: {
+    fontSize: '14px',
+    fontWeight: 600,
+    letterSpacing: 0,
+    lineHeight: '22px',
+    color: colors.primary,
+    marginTop: '2rem',
+    marginBottom: '0.3rem'
+  }
+});
+
+interface InstitutionEditState {
+  apiErrorMsg: string;
+  isAddInstitution: boolean;
+  institution: Institution;
+  invalidEmailAddress: boolean;
+  invalidEmailAddressMsg: string;
+  invalidEmailDomain: boolean;
+  invalidEmailDomainsMsg: string;
+  showOtherInstitution: boolean;
+  showBackButtonWarning: boolean;
+  showApiError: boolean;
+}
+
+let title = 'Add new Institution';
+let institutionToEdit;
+
+export class AdminInstitutionEditImpl extends React.Component<UrlParamsProps, InstitutionEditState> {
+  constructor(props) {
+    super(props);
+    this.state = {
+      apiErrorMsg: '',
+      isAddInstitution: true,
+      institution: {
+        shortName: '',
+        displayName: '',
+        organizationTypeEnum: null
+      },
+      invalidEmailAddress: false,
+      invalidEmailAddressMsg: '',
+      invalidEmailDomain: false,
+      invalidEmailDomainsMsg: '',
+      showOtherInstitution: false,
+      showBackButtonWarning: false,
+      showApiError: false
+    };
+  }
+
+  async componentDidMount() {
+    // If institution short Name is passed in the URL get the institution details
+    if (this.props.urlParams.institutionId) {
+      institutionToEdit = await institutionApi().getInstitution(this.props.urlParams.institutionId);
+      title = institutionToEdit.displayName;
+      this.setState({
+        isAddInstitution: false,
+        institution: institutionToEdit,
+        showOtherInstitution: institutionToEdit.organizationTypeEnum === OrganizationType.OTHER
+      });
+    } else {
+      title = 'Add new Institution';
+      this.setState({isAddInstitution: true});
+    }
+
+  }
+
+  get institutionTypeOptions() {
+    const options = [
+      {label: 'Industry', value: OrganizationType.INDUSTRY},
+      {label: 'Academic Research Institution', value: OrganizationType.ACADEMICRESEARCHINSTITUTION},
+      {label: 'Educational Institution', value: OrganizationType.EDUCATIONALINSTITUTION},
+      {label: 'Health Center non profit', value: OrganizationType.HEALTHCENTERNONPROFIT},
+      {label: 'Other', value: OrganizationType.OTHER}
+    ];
+    return options;
+  }
+
+  get institutionAgreementTypeOptions() {
+    const options = [
+      {label: 'Master', value: DuaType.MASTER},
+      {label: 'Individual', value: DuaType.RESTRICTED}
+    ];
+    return options;
+  }
+
+  // Filter out empty line or empty email addresses like <email1>,,<email2>
+  // Confirm each email is a valid email using validate.js
+  validateEmailAddresses() {
+    const invalidEmailAddress = [];
+    const {emailAddresses} = this.state.institution;
+    this.state.institution.emailAddresses = emailAddresses.filter(
```

Comment:
is this a safe way to set state?  We should probably use a local variable here and call setState() to update this list when we're done - or maybe it would be better not to update the state of `emailAddresses` at all?